### PR TITLE
Merge sort (Python, sorting)

### LINKED ISSUE #6: merge_sort
Issue for coding challenge. Let's take the first one not to hard.

  - milestone: coding challenges, sorting, merge_sort
  - labels: sorting
  - 
https://github.com/MIT-Emerging-Talent/2024-group-02-collaboration-practice/tree/main/src/sorting/python/merge_sort

### PR DESCRIPTION
**Pull request by tvsirius**
-->

- [x] PR's title includes the challenge's name and language
- [x] PR has short and clear description of the challenge
- [x] PR has appropriate labels and milestones for easy identification
- [x] PR it is assigned to the owner
- [x] reviewers are assigned
- [x] the PR contributes only one focused change
- [x] the branch is up to date with `main`/`master`
- [x] the code works when pulled and run locally
- [x] all conflicts are resolved (if any)
- [x] It is linked to an issue in the appropriate column of the project board
- [x] feedback is addressed (if any, and if it is appropriate feedback.)

## README Documentation

- [x] The solution is documented in `/src/README.md`
- [x] The markdown source is formatted
- [x] Spelling and grammar is correct in all text
- [x] The markdown looks correct when you preview the file
- [x] All links and images work
- [x] The README documents the solution's behavior, strategy(ies) and implementation(s)
- [x] There are use cases in code blocks to illustrate the function's behavior

## Python Files

- [x] There is a module header
- [x] There is a module docstring
- [x] File names are in snake_case
- [x] Test files are named `test_<module_name>.py`

## Function Docstring

- [x] Behavior description
- [x] Parameter description
- [x] Return value description
- [x] Include any assumptions
- [x] Include 3 or more (passing!) doctests
- [ ] Include 1-2 use cases (if necessary)

## Function Implementation

- [ ] The solution is not is not copied from [djeada](https://github.com/djeada/Algorithms-And-Data-Structures)
- [x] The code is auto-formatted
- [x] The code has no (reasonable) linting mistakes
- [x] Variables are named with snake_case
- [x] The function has a clear and helpful name
- [x] The file's name matches the function name
- [x] The code follows the strategy as simply as possible
- [x] Variable names are clear and helpful
- [x] Comments explain the strategy (if necessary)
- [x] There are type annotations
- [x] (challenge) The code includes defensive assertions

## Unit Test Suite

- [x] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Each unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [x] Only one assertion
- [x] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases
- [ ] Test suite includes black-box and glass-box tests

### CODE REVIEW

File: src/sorting/python/merge_sort/src/merge_sort.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    [2,3]
+
+    >>> merge_sort([1,5,10,2,22],ascending = False)
+    [22,10,5,2,1]
+    
+    
+    """
+
+    if len(list_to_sort) == 0:
+        return []
```
Comment (JumongEric) [inline src/sorting/python/merge_sort/src/merge_sort.py:37]: Recursion simplify the work. I like this.

File: src/sorting/python/merge_sort/src/merge_sort.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on 17 01 24
+
+@author: tvsirius
+"""
+
+
+def merge_sort(list_to_sort: list[any], ascending: bool = True) -> list[any]:
```
Comment (JumongEric) [inline src/sorting/python/merge_sort/src/merge_sort.py:10]: I wasn't familiar with this function annotation, but thanks now I know

File: src/sorting/python/merge_sort/src/merge_sort.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    if len(list_to_sort) == 0:
+        return []
+
+    assert isinstance(list_to_sort, (list, tuple)
+                      ), "list_to_sort must be list or tuple"
+    # I will not test all items compatibility,
+    # I will assert first element for allowing comparison methods (< > <= >=)
+
+    if not (hasattr(list_to_sort[0], '__lt__') and
+            hasattr(list_to_sort[0], '__le__') and
```
Comment (JumongEric) [inline src/sorting/python/merge_sort/src/merge_sort.py]: Even though I can see what this snippet does, I am confused about how it works. I used to think that we used these methods in defining classes.

File: src/sorting/python/merge_sort/src/merge_sort.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        mid_point = len(list_to_sort)//2
+        # Zero case
+        if len(list_to_sort) < 2:
+            return list_to_sort
+
+        # Recursion
+        return helper_merge(helper_sort_recursion(list_to_sort[0:mid_point]),
+                            helper_sort_recursion(list_to_sort[mid_point:]))
+
+    return helper_sort_recursion(list_to_sort)
```
Comment (JumongEric) [inline src/sorting/python/merge_sort/src/merge_sort.py:95]: Seems to be fine.

File: src/sorting/python/merge_sort/src/merge_sort.py (hunk 3 above)
Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: It checking attributes, but it always will be false. 
@tvsirius, are you sure you need this check? Can you give example of type without such attributes? I seems cant find any

File: src/sorting/python/merge_sort/src/merge_sort.py (hunk 2 above)
Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py:10]: Check constraints.md file. It has some explanation

File: src/sorting/python/merge_sort/src/merge_sort.py (hunk 3 above)
Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Yes, you cannot compare strings.
There can be custom classes with implementations of ><>=<=, and can be without. So I want sort to work with those, who can

Other thinf this is a check for first elements, and others are checked with try-except later, so I don't know if it is a good way, but it works

File: src/sorting/python/merge_sort/src/merge_sort.py (hunk 4 above)
Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py:95]: So, you can aprove this PR to merge?

File: src/sorting/python/merge_sort/src/merge_sort.py (hunk 3 above)
Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Check it out:

print("bcd">"abc")
print("abc">"bcd")

As for custom classes. Every clas is a child of Object class, and object class have all these methods. Custom class maybe not implemented these methods but already has them by default. 

Looks like all types in Python is classes. Even primitives as int, bool, float. I have no idea where function hasattr can be used. Looks like it obsolete, and exist fo some comparability with old code. 

Can you rework those checks? You can take idea of how I implemented it in my pull request

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: @colevandersWands

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: But you implemnted it wrong, because you cannot sort (int, float) list

I think i can remove this first check, because I have do it with 
try - except block inside code

File: src/sorting/python/merge_sort/src/merge_sort.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                        j += 1
+                else:
+                    # Descending
+                    if list1[i] >= list2[j]:
+                        result_list.append(list1[i])
+                        i += 1
+                    else:
+                        result_list.append(list2[j])
+                        j += 1
+            except TypeError as exc:
```
Comment (tvsirius) [inline src/sorting/python/merge_sort/src/merge_sort.py:71]: @Vlad421

File: src/sorting/python/merge_sort/src/merge_sort.py (hunk 3 above)
Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Why in the world we need to sort mixed int float type? Where did you get this idea?

Let's think theoretically. If someone works with dataset. It has mixed data. It his responsibility to clean it and prepare. To let's say yours sorting algorithm. Not your.
Because let's pretend something goes wrong and you (some python behavior. Couldnt submit problem set in 02 course because on my PC it gave exact output grader expected, but grader because of Python versions or NumPy versions gave results with .00000000000000000001 at every position) cast in some part of code float to int, in output it would be dataset with errors. Not you problem. Whoever will use have to prepare data.

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: I realy don't know, maybe we all have the point. 
Just built in .sort() and sorted() functions in python, as i know do sort of everything,
and even your code will do it, if it goes over assertions

I have the same issue with PS, as I remember.  I have used  round, multiplication by 1000 and rounding again, just to pass the grader

ok, then, we can ask @colevandersWands about this, and leave our solutions as they are

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: You cannot compare dicst for example
>>> {"f":"f"}<{} 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'dict' and 'dict'
>>>

**Comment by tvsirius** [comment]
Commited last updates, first check of list_to_sort[0] is really not needed

File: src/sorting/python/merge_sort/src/merge_sort.py (hunk 3 above)
Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: You not supposed to compare two dic at all or lists or tuples. It doesn't have any sense.

You may compare entire collections, but you need to do some math with it's elements before and get some results from it for comparison

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Yes, of course. 
But it's the defensive programming thing we are talking about

Comment (colevandersWands) [reply src/sorting/python/merge_sort/src/merge_sort.py]: I would say this is a case where you can decide how you want it to work.  The original tests didn't specify so you get to decide.  If you want a function that only compares the same type, you can have a defensive check to make sure they're the same type (and a matching unit test).  If you want it to only work for eg. ints you can do a defensive type check (and a matching unit test).

What ever you do, just agree to it in code review and document this behavior with type signatures, docstrings and unit tests.

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Yes, that's what I also thinks. I think for this challenge it's ok to go different ways, just we should document it. So I will update README.md for this challenge.
For the real project we should be more specific, but there we will have the context, that will set our constrains

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Oh, it's already documented

**Comment by tvsirius** [comment]
So this PR can be approved

File: src/sorting/python/merge_sort/src/merge_sort.py (hunk 3 above)
Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: > I would say this is a case where you can decide how you want it to work. The original tests didn't specify so you get to decide. If you want a function that only compares the same type, you can have a defensive check to make sure they're the same type (and a matching unit test). If you want it to only work for eg. ints you can do a defensive type check (and a matching unit test).
> 
> What ever you do, just agree to it in code review and document this behavior with type signatures, docstrings and unit tests.
@colevandersWands 
Yes. But point is that piece of code will be never reached because all data in Python has that attributes. They might be not implemented but hasattr(arr[i]) will always be true and making it inside if not block will make it false which means this block will never be entered by interpreter.

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: There are data types without '__lt__' '__le__'... attributes

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: > There are data types without '**lt**' '**le**'... attributes

Do you have examples?

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Check this out

isinstance(1, object) # True
isinstance('Hello World', object) # True
isinstance(int, object) # True
isinstance(object, object) # True
isinstance(type, object) # True

https://stackoverflow.com/a/49143578

Every type in Python is a subclass of Object. Object have all these attributes, and children classes as well

Comment (colevandersWands) [reply src/sorting/python/merge_sort/src/merge_sort.py]: @Vlad421  good point, I wasn't thinking of it at that level.  In that case it is extra code making the solution harder to understand

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: What I propose is replace attributes test for type test. To exclude mixed type inside data collection, and exclude everything except tuples and lists. 
We could sort dictionaries but for dictionaries we need one more parameter for function to specify sorting by keys or values. Which is redundant it current scope. Even in Python for sorting dict you create 2 lists from keys and values, sort them simultaneously and then combine to dict

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: > Check this out
> 
> isinstance(1, object) # True isinstance('Hello World', object) # True isinstance(int, object) # True isinstance(object, object) # True isinstance(type, object) # True
> 
> https://stackoverflow.com/a/49143578
> 
> Every type in Python is a subclass of Object. Object have all these attributes, and children classes as well

It was about this specific attiributes:
    if not (hasattr(list_to_sort[0], '__lt__') and
            hasattr(list_to_sort[0], '__le__') and ...
you are showing that everything is an instance of object and it's not related here

There are objects that cannot be compared with lt le gt ge methods (i.e. dict, and many custom classes)


As we said before it's a matter of convention, for here we don't have strict rules. I pointed out in README my sort can sort everything that can be compared, in your sort you even prohibit sorting mixed list of int and float
I think we are both fine
I can do of course a strict version of my sort if you like


For now I removed the hasattr(list_to_sort[0], '__lt__') check beacuse I am checking it directly with try except block


When we will write function for DS project we will have context that will say us how we should make functions work

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: >There is onbject that cannot be compared, but elements can. We comparing elements of collections not collections. List, tuples is collections too
>(i.e. dict, and many custom classes)

Dict is collection. It can't be compared. You have to understand it.
Custom classes. Every class is subclass of Object class. Even if you didn't tell it explicitly

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: List and tuples is collections too. We are comparing elements inside collection not collections between each other

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: @tvsirius 
https://docs.python.org/3/reference/datamodel.html#basic-customization

Docs for Object class

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: >>> Custom classes. Every class is subclass of Object class. Even if you didn't tell it explicitly
To this I agree

>>> We are comparing elements inside collection not collections between each other

Yes, so the point was what data types cannot be compared.

our input list to sort should be list or tuple
elements of this list must allow comapison. if we have a list of int,chars,floats,string even list of lists,sets of tuple and any custom type that allows comparision it will sort it

if elemets of list cannot be compared, such as if we pass a list of dicts, or some user type it will return a TypeErorr

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Also some code snippet for not believers 😁


class Test:
    def __init__(self):
        print("Hello")
        
        
a = Test()

print(isinstance(a, object))
print(isinstance(5, object ))
print(isinstance('a', object))

Man.. Python is fully OOP language. With unfortunately not best OOP implementation 😂

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: > Also some code snippet for not believers 😁

It seems you got me wrong somehow, i was not aguing that everything is an object
read my previous comment

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Code in if block is unreachable. It will never enters inside that block. And never throw an exception. This is a point. You need to think about it. It actually great opportunity to learn something new. 

Look at you code and try to play it inside you head. Step by step. You will see

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: To simplify it look like this:
if not true:  #false all time 
....Exception

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: yes, it seems you right, maybe every object has this attributes, but for non comparatable they return an errors
```
def test_comparison(obj_test):
    return (hasattr(obj_test, '__lt__') and
            hasattr(obj_test, '__le__') and
            hasattr(obj_test, '__gt__') and
            hasattr(obj_test, '__ge__'))


print(test_comparison('1'))
print(test_comparison(3))
print(test_comparison(test_comparison))
print(test_comparison(print))
print(test_comparison(object))
print(test_comparison({"a":"1","v":3}))
print(test_comparison('2f'))
print(test_comparison([1, 2,4]))
```
True all the time
Thank you but i didn't get your point from the start

File: src/sorting/python/merge_sort/src/merge_sort.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    def helper_merge(list1, list2) -> list:
+        """Helper function for merging two lists, based on ascending parameter
+        """
+        i = j = 0
+        list1_len = len(list1)
+        list2_len = len(list2)
+        result_list = []
+        # while both list are not empty
+        while i < list1_len and j < list2_len:
+            try:
```
Comment (tvsirius) [inline src/sorting/python/merge_sort/src/merge_sort.py:53]: but this code works @Vlad421

File: src/sorting/python/merge_sort/src/merge_sort.py (hunk 3 above)
Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: And also this code has been already removed

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: > but this code has been already removed

Sorry. Didn't see it. But maybe better not to comment but remove? Or change it for mixed type check?

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: > True all the time


Exactly. True all the time. And you explicit 
If **not**. Pay attention you have not statement which reverse Boolean value to opposite. Turns it to False. Logical if enters inside only if expression equal True . Which means it will never enter that block to throw exception. And made no use for that check. Because it will never happen. This is what I tell about

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Ok, now i got what you mean

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Dunno why you vere provings everything is an object )

Comment (delvor) [reply src/sorting/python/merge_sort/src/merge_sort.py]: > Dunno why you vere provings everything is an object )

Because it how Python is made. Funny language.

Comment (tvsirius) [reply src/sorting/python/merge_sort/src/merge_sort.py]: Yeah
I like it a lot

**Review by delvor — APPROVED** [review]
looks good
